File: packages/engine/src/cores/famibob/cartridge.js
import { Mirroring, Region } from '.'

export class Cartridge {
  /**
   * @param {string} filename
   * @param {number[]} romData
   */
  constructor(filename, romData) {
    if (
      romData[0] === 'N'.charCodeAt(0) &&
      romData[1] === 'E'.charCodeAt(0) &&
      romData[2] === 'S'.charCodeAt(0) &&
      romData[3] === 0x1a
    ) {
      if ((romData[7] & 3) !== 0) throw 'Unsupported system'

      console.log('Parsing NES ROM...')
      console.log(`Filename: ${filename}`)
      console.log(`ROM size: ${romData.length}`)

      this.prgRomSize = 16 * 1024 * romData[4]
      this.chrRomSize = 8 * 1024 * romData[5]
      this.mapperId = (romData[6] >>> 4) | (romData[7] & 0xf0)
      this.trainer = (romData[6] & 4) > 0
      this.battery = (romData[6] & 2) > 0

      if ((romData[6] & 8) > 0) this.mirroring = Mirroring.FOUR_SCREEN
      else if ((romData[6] & 1) > 0) this.mirroring = Mirroring.VERTICAL
      else this.mirroring = Mirroring.HORIZONTAL

      if ((romData[7] & 0xc) === 8) {
        console.log('NES 2.0 file format.')
        this.prgRamSize = 64 << (romData[10] & 0xf || 7)
        this.chrRamSize = 64 << (romData[11] & 0xf)
        this.region = (romData[12] & 3) !== 1 ? Region.NTSC : Region.PAL
      } else {
        console.log('iNES file format.')
        if (['[pal]', '(pal)', '[e]', '(e)', '[europe]', '(europe)'].includes(filename.toLowerCase())) {
          console.log('Region set to PAL based on the file name.')
          this.region = Region.PAL
        } else this.region = Region.NTSC
        this.prgRamSize = 8 * 1024
        this.chrRamSize = 0
      }

      console.table(this)

      const prgFrom = this.trainer ? 528 : 16
      this.prgData = romData.slice(prgFrom, prgFrom + this.prgRomSize)
      this.chrData = romData.slice(romData.length - this.chrRomSize)
      this.trainerData = this.trainer ? romData.slice(16, 528) : []

      // if (this.mapperId === 0 && this.chrData.length === 0 && this.prgData.length === 16 * 1024) {
      //   console.log('No CHR ROM found. Fixing ranges...')
      //   this.chrData = romData.slice(romData.length - 8 * 1024)
      //   this.prgData = romData.slice(prgFrom, prgFrom + 8 * 1024)
      //   this.prgData = [...this.prgData, ...this.prgData]
      // }

      console.log(`Region: ${Object.keys(Region).find((d) => Region[d] === this.region)}`)
      console.log(`Mirroring: ${Object.keys(Mirroring).find((d) => Mirroring[d] === this.mirroring)}`)
      console.log('Loaded PRG ROM size: ', this.prgData.length)
      console.log('Loaded CHR ROM size: ', this.chrData.length)
      console.log('Loaded trainer size: ', this.trainerData.length)
    } else {
      throw 'Invalid file'
    }
  }
}

File: packages/engine/src/cores/famibob/cartridge.zig
const std = @import("std");
const Region = @import("../core.zig").Region;
const Mirroring = @import("famibob.zig").Mirroring;

pub const Cartridge = struct {
    allocator: std.mem.Allocator,
    prg_data: []u8,
    chr_data: []u8,
    trainer_data: ?[]u8 = null,

    crc: u32 = 0,
    mapper_id: u8 = 0,
    prg_rom_size: usize = 0,
    chr_rom_size: usize = 0,
    prg_ram_size: usize = 0,
    chr_ram_size: usize = 0,
    trainer: bool = false,
    battery: bool = false,
    region: Region = .ntsc,
    mirroring: Mirroring = .horizontal,

    pub fn init(allocator: std.mem.Allocator, rom_data: []const u8) !*Cartridge {
        if (rom_data[0] == 'N' and rom_data[1] == 'E' and rom_data[2] == 'S' and rom_data[3] == 0x1a) {
            if ((rom_data[7] & 3) != 0) {
                return error.UnsupportedSystem;
            }

            std.debug.print("Parsing NES ROM...\n", .{});
            std.debug.print("ROM size: {d}\n", .{rom_data.len});

            var instance = try allocator.create(Cartridge);

            instance.* = .{
                .allocator = allocator,
                .prg_data = undefined,
                .chr_data = undefined,
                .crc = std.hash.Crc32.hash(rom_data),
                .prg_rom_size = 16 * 1024 * @as(usize, rom_data[4]),
                .chr_rom_size = 8 * 1024 * @as(usize, rom_data[5]),
                .mapper_id = (rom_data[6] >> 4) | (rom_data[7] & 0xf0),
                .trainer = (rom_data[6] & 4) > 0,
                .battery = (rom_data[6] & 2) > 0,
            };

            if ((rom_data[6] & 8) > 0) {
                instance.mirroring = .four_screen;
            } else if ((rom_data[6] & 1) > 0) {
                instance.mirroring = .vertical;
            } else {
                instance.mirroring = .horizontal;
            }

            if ((rom_data[7] & 0xc) == 8) {
                std.debug.print("NES 2.0 file format.\n", .{});
                instance.prg_ram_size = if (rom_data[10] & 0xf > 0) @as(usize, 64) << @as(u4, @truncate(rom_data[10] & 0xf)) else 64 << 7;
                instance.chr_ram_size = @as(usize, 64) << @as(u4, @truncate(rom_data[11] & 0xf));
                instance.region = if ((rom_data[12] & 3) != 1) .ntsc else .pal;
            } else {
                std.debug.print("iNES file format\n", .{});
                instance.prg_ram_size = 8 * 1024;
                instance.chr_ram_size = 0;
                instance.region = .ntsc;
            }

            const prg_from: usize = if (instance.trainer) 528 else 16;
            instance.prg_data = try allocator.alloc(u8, instance.prg_rom_size);
            if (instance.prg_data.len > 0) {
                @memcpy(instance.prg_data, rom_data[prg_from .. prg_from + instance.prg_rom_size]);
            }

            instance.chr_data = try allocator.alloc(u8, instance.chr_rom_size);
            if (instance.chr_data.len > 0) {
                @memcpy(instance.chr_data, rom_data[rom_data.len - instance.chr_rom_size ..]);
            }

            if (instance.trainer) {
                instance.trainer_data = try allocator.alloc(u8, 512);
                @memcpy(instance.trainer_data.?, rom_data[16..528]);
            }

            std.debug.print("Mapper ID: {d}\n", .{instance.mapper_id});
            std.debug.print("Region: {s}\n", .{std.enums.tagName(Region, instance.region).?});
            std.debug.print("Mirroring: {s}\n", .{std.enums.tagName(Mirroring, instance.mirroring).?});
            std.debug.print("Battery: {any}\n", .{instance.battery});
            std.debug.print("Trainer data: {any}\n", .{instance.trainer});
            std.debug.print("PRG ROM size: {d}\n", .{instance.prg_data.len});
            std.debug.print("CHR ROM size: {d}\n", .{instance.chr_data.len});
            std.debug.print("PRG RAM size: {d}\n", .{instance.prg_ram_size});
            std.debug.print("CHR RAM size: {d}\n", .{instance.chr_ram_size});

            return instance;
        } else {
            return error.UnsupportedFile;
        }
    }

    pub fn deinit(self: *Cartridge) void {
        self.allocator.free(self.prg_data);
        self.allocator.free(self.chr_data);
        if (self.trainer_data) |t| self.allocator.free(t);
        self.allocator.destroy(self);
    }
};
